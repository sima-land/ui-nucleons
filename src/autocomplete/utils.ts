import type { SyntheticEvent } from 'react';

/**
 * Наивная реализация синтетического события.
 * Не рекомендуется к использованию за пределами библиотеки.
 */
export class NaiveSyntheticEvent<T extends Element = Element> implements SyntheticEvent<T> {
  nativeEvent: Event;
  currentTarget: EventTarget & T;
  target: EventTarget & T;
  bubbles: boolean;
  cancelable: boolean;
  defaultPrevented: boolean;
  eventPhase: number;
  isTrusted: boolean;
  timeStamp: number;
  type: string;

  /** @inheritdoc */
  constructor(nativeEvent: Event, target: T) {
    this.target = target;
    this.currentTarget = target;
    this.nativeEvent = nativeEvent;
    this.type = nativeEvent.type;
    this.timeStamp = nativeEvent.timeStamp;
    this.bubbles = nativeEvent.bubbles;
    this.cancelable = nativeEvent.cancelable;
    this.defaultPrevented = nativeEvent.defaultPrevented;
    this.eventPhase = nativeEvent.eventPhase;
    this.isTrusted = nativeEvent.isTrusted;
  }

  /** @inheritdoc */
  persist() {
    return void 0;
  }

  /** @inheritdoc */
  stopPropagation() {
    this.nativeEvent.stopPropagation();
  }

  /** @inheritdoc */
  isPropagationStopped() {
    return false;
  }

  /** @inheritdoc */
  preventDefault() {
    this.nativeEvent.preventDefault();
    this.defaultPrevented = this.nativeEvent.defaultPrevented;
  }

  /** @inheritdoc */
  isDefaultPrevented() {
    return this.defaultPrevented;
  }
}

/**
 * Вызывает переданную функцию только если переданное значение не является null или undefined.
 * @param value Значение.
 * @param callback Функция.
 * @return Результат функции или undefined.
 */
export function when<A, B>(value: A, callback: (value: NonNullable<A>) => B): B | undefined {
  return value !== null && value !== undefined ? callback(value) : undefined;
}
