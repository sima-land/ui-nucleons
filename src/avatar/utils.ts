/**
 * Формирует монограмму из переданной строки.
 * @param value Переданная строка.
 * @return Монограмма.
 */
export function getMonogram(value: any): string {
  if (typeof value === 'string') {
    return value
      .split(/\s+/g)
      .slice(0, 2)
      .map(s => s[0])
      .reverse()
      .join('')
      .toUpperCase();
  }

  return '';
}
