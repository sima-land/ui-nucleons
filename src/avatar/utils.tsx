import type { SVGAttributes } from 'react';
import type { AvatarProps } from './types';
import PersonSVG from '@sima-land/ui-quarks/icons/24x24/Stroked/Person';
import classNames from 'classnames';
import styles from './utils.module.scss';

const USER_AVATAR_COLOR_TOKENS: ReadonlyArray<string> = [
  '#eb8585',
  '#fda09b',
  '#f49bb1',
  '#ffd785',
  '#ffbb85',
  '#89d7c7',
  '#85e5ac',
  '#b5ef90',
  '#93c4ee',
  '#85caf5',
  '#85ddea',
  '#b3c1c7',
  '#dc9bd5',
  '#b4a8f0',
  '#c999e9',
];

/**
 * Возвращает свойства для аватара пользователя.
 * @param payload Данные пользователя.
 * @param customProps Дополнительные свойства аватара.
 * @return Свойства.
 */
export function getUserAvatarProps(
  {
    id,
    name,
    image,
  }: {
    id?: number;
    name?: string;
    image?: string;
  },
  customProps?: AvatarProps,
): AvatarProps {
  return {
    src: image,
    children: getNameInitials(name) || <PersonSVG {...getAvatarIconProps()} />,
    ...customProps,
    className: classNames(styles.user, customProps?.className),
    style: {
      '--avatar-color': id ? getUserAvatarColor(`${id}`) : undefined,
      ...customProps?.style,
    },
  };
}

/**
 * Возвращает цвет аватара текущего пользователя.
 * @param identity Срока, идентифицирующая пользователя.
 * @return Цвет.
 */
export function getUserAvatarColor(identity: string): string {
  let total = 0;

  for (let i = 0; i < identity.length; i++) {
    total += identity.charCodeAt(i);
  }

  const list = USER_AVATAR_COLOR_TOKENS;
  const hash = total % list.length;

  return list[hash];
}

/**
 * Формирует инициалы имени.
 * @param value Строка.
 * @return Инициалы.
 */
export function getNameInitials(value: any): string {
  if (typeof value === 'string') {
    return (value[0] ?? '').toUpperCase();
  }

  return '';
}

/**
 * Возвращает свойства для иконки внутри аватара.
 * @return Свойства.
 */
export function getAvatarIconProps(): SVGAttributes<SVGSVGElement> {
  return {
    className: styles.icon,
  };
}
