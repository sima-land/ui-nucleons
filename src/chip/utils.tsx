import type { ChipProps } from './types';
import { ChipIconButton } from './chip';
import CrossSVG from '@sima-land/ui-quarks/icons/16x16/Stroked/Cross';

/**
 * Возвращает свойства для чипа, добавляющие ему кнопку удаления.
 * @param options Опции.
 * @return Свойства.
 */
export function getDeletableChipProps({ onDelete }: { onDelete?: VoidFunction } = {}): ChipProps {
  return {
    padding: 'start',
    adornmentGutter: 'unset',
    endAdornment: (
      <ChipIconButton type='button' onClick={onDelete}>
        <CrossSVG fill='currentColor' />
      </ChipIconButton>
    ),
  };
}
