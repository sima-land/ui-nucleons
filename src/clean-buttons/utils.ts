import type { CleanButtonProps } from './types';
import { createContext } from 'react';

/**
 * Контекст установки размера группы прозрачных кнопок CleanGroup.
 * Размер заданный непосредственно компоненту CleanGroup имеет приоритет над контекстом.
 */
export const CleanButtonContext = createContext<Pick<CleanButtonProps, 'size' | 'className'>>({});
