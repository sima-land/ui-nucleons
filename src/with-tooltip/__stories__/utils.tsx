import { HTMLAttributes, ReactNode, useEffect, useRef, useState } from 'react';
import { WithTooltip } from '..';
import boundsOf from '../../helpers/bounds-of';
import on from '../../helpers/on';
import { Link } from '../../link';

export const SHORT_TEXT = 'Short tooltip text.';

export const LONG_TEXT = [
  'Тормоза делятся на 3 основных вида: барабанный (ножной), ободной и дисковый.',
  'Преимущество барабанного тормоза заключается в простоте обслуживания.',
  'Такой тормоз долговечен, потому что механизм закрыт, и в него попадает мало грязи.',
  'Он не изнашивает обод. Ободной тормоз отличается наименьшим весом.',
  'В сравнении с барабанным, обеспечивает более плавное торможение.',
  'Дисковый тормоз эффективен в любых погодных условиях, потому что на механизм не попадает грязь с ободьев.',
  'Обеспечивает мощное торможение, подходит для экстремальной езды по бездорожью. Не изнашивает обод. ',
]
  .join(' ')
  .repeat(2);

export function DismissTest() {
  return (
    <WithTooltip tooltip='Hello, world!'>
      {(ref, toggle) => (
        <Link ref={ref as any} pseudo onClick={() => toggle(true)}>
          [Toggle tooltip]
        </Link>
      )}
    </WithTooltip>
  );
}

export function LinkWithTooltip({ tooltip = SHORT_TEXT }: { tooltip?: ReactNode }) {
  return (
    <WithTooltip tooltip={tooltip}>
      {(ref, toggle) => (
        <Link ref={ref as any} pseudo onClick={() => toggle(true)}>
          Show tooltip
        </Link>
      )}
    </WithTooltip>
  );
}

export function SquareWithTooltip() {
  return (
    <WithTooltip tooltip='There is tooltip'>
      {(ref, toggle, shown) => (
        <div
          ref={ref as any}
          style={{
            width: 24,
            height: 24,
            margin: 12,
            background: shown ? '#aaa' : '#ddd',
            flexShrink: 0,
            cursor: 'pointer',
          }}
          onClick={() => toggle(true)}
        />
      )}
    </WithTooltip>
  );
}

export function PositioningTest({
  tooltip,
  containerProps,
}: {
  tooltip: ReactNode;
  containerProps?: HTMLAttributes<HTMLDivElement>;
}) {
  const { draggableRef, movedRef } = useDraggable();

  return (
    <div
      style={{
        width: 2000,
        height: 1000,
        border: '10px solid #aaa',
      }}
      {...containerProps}
    >
      <div ref={draggableRef as any} style={{ position: 'absolute' }}>
        <WithTooltip tooltip={tooltip}>
          {(ref, toggle) => (
            <Link
              ref={ref as any}
              onClick={() => {
                !movedRef.current && toggle(true);
                movedRef.current = false;
              }}
              pseudo
            >
              [draggable toggle]
            </Link>
          )}
        </WithTooltip>
      </div>
    </div>
  );
}

export function useDraggable() {
  const movedRef = useRef(false);
  const capturedRef = useRef(false);
  const captureOffsetRef = useRef({ x: 0, y: 0 });

  const draggableRef = useRef<HTMLElement>();
  const [, setCount] = useState(0);

  useEffect(() => {
    const list = [
      on<MouseEvent>(
        window,
        'mousedown',
        event => {
          if (
            event.target === draggableRef.current ||
            (draggableRef.current as any).contains((event as any).target)
          ) {
            event.preventDefault();
            captureOffsetRef.current.x =
              (boundsOf(draggableRef.current) as any).left - event.clientX;
            captureOffsetRef.current.y =
              (boundsOf(draggableRef.current) as any).top - event.clientY;
            capturedRef.current = true;
          }
        },
        { capture: true },
      ),
      on<MouseEvent>(window, 'mousemove', ({ clientX, clientY }) => {
        if (capturedRef.current) {
          const { current: captureOffset } = captureOffsetRef;

          movedRef.current = true;
          (draggableRef.current as any).style.top = `${
            window.pageYOffset + clientY + captureOffset.y
          }px`;
          (draggableRef.current as any).style.left = `${
            window.pageXOffset + clientX + captureOffset.x
          }px`;
          setCount(Math.random()); // для пересчета позиции
        }
      }),
      on(window, 'mouseup', () => {
        capturedRef.current = false;
        setCount(Math.random()); // для пересчета позиции
      }),
    ];

    return () => list.forEach(fn => fn());
  }, []);

  return { draggableRef, movedRef };
}
