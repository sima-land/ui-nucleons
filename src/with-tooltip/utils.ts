import { boundsOf } from '../helpers/bounds-of';
import { getScrollParent } from '../helpers/get-scroll-parent';
import { findOffsetParent } from '../helpers/find-offset-parent';

type DOMRectLike = Record<(typeof rectKeys)[number], number>;

type CanPlaceFn = (holder: HTMLElement, tooltipEl: HTMLElement, area: DOMRectLike) => boolean;

const SPACES = {
  correction: 24, // сумма toHolder и toAreas
  toHolder: 8,
  toArea: 16,
} as const;

const rectKeys = ['top', 'left', 'bottom', 'right', 'width', 'height'] as const;

/**
 * Преобразует css-значение размера в число.
 * @param cssValue Значение, например "12.23px".
 * @return Число. Неожиданно, правда?
 */
const asNumber = (cssValue: string) => parseFloat(cssValue.replace(/[A-z]/g, '')) || 0;

/**
 * Складывает все аргументы.
 * @param  numbers Числа.
 * @return Число. Неожиданно, правда?
 */
const sum = (...numbers: number[]) => numbers.reduce((a, b) => a + b);

export const CanPlace: Record<'onRight' | 'onLeft' | 'onBottom' | 'onTop', CanPlaceFn> = {
  onRight: (holderEl, tooltipEl, area) =>
    area.right - boundsOf(holderEl).right - SPACES.correction > boundsOf(tooltipEl).width,
  onLeft: (holderEl, tooltipEl, area) =>
    boundsOf(holderEl).left - SPACES.correction - area.left > boundsOf(tooltipEl).width,
  onBottom: (holderEl, tooltipEl, area) =>
    boundsOf(tooltipEl).height < area.bottom - boundsOf(holderEl).bottom - SPACES.correction,
  onTop: (holderEl, tooltipEl, area) =>
    boundsOf(tooltipEl).height < boundsOf(holderEl).top - area.top - SPACES.correction,
};

export const PlaceOffset = {
  reset: (element: HTMLElement) => {
    element.style.top = '0';
    element.style.left = '0';
    element.style.bottom = '';
    element.style.right = '';
    element.style.maxHeight = '';
  },
  onRight: (tooltipEl: HTMLElement, holderEl: HTMLElement) => {
    const correction = getOriginCorrection(tooltipEl);

    return sum(SPACES.toHolder, boundsOf(holderEl).right, correction.x);
  },
  onLeft: (tooltipEl: HTMLElement, holderEl: HTMLElement) => {
    const correction = getOriginCorrection(tooltipEl);

    return sum(boundsOf(holderEl).left, -boundsOf(tooltipEl).width, -SPACES.toHolder, correction.x);
  },
  onBottom: (tooltipEl: HTMLElement, holderEl: HTMLElement) => {
    const correction = getOriginCorrection(tooltipEl);

    return sum(boundsOf(holderEl).bottom, SPACES.toHolder, correction.y);
  },
  onTop: (tooltipEl: HTMLElement, holderEl: HTMLElement) => {
    const correction = getOriginCorrection(tooltipEl);

    return sum(boundsOf(holderEl).top, -boundsOf(tooltipEl).height, -SPACES.toHolder, correction.y);
  },
};

export const CorrectOffset = {
  vertically: (tooltipEl: HTMLElement, holderEl: HTMLElement, area: DOMRectLike) => {
    const tooltipBounds = boundsOf(tooltipEl) as DOMRect;
    const holderBounds = boundsOf(holderEl) as DOMRect;
    const correction = getOriginCorrection(tooltipEl);

    let result;

    if (area.bottom - holderBounds.bottom > tooltipBounds.height + SPACES.correction) {
      result = correction.y + holderBounds.top;
    } else if (holderBounds.bottom - area.top > tooltipBounds.height + SPACES.correction) {
      result = correction.y + holderBounds.bottom - tooltipBounds.height;
    } else {
      result = correction.y + SPACES.toArea;
    }

    return result;
  },
  horizontally: (tooltipEl: HTMLElement, area: DOMRectLike) => {
    const correction = getOriginCorrection(tooltipEl);

    return correction.x + area.left + SPACES.toArea;
  },
};

/**
 * Позиционирует элемент подсказки относительно "держащего" элемента.
 * @param tooltipEl Элемент подсказки.
 * @param holderEl Держащий элемент.
 * @param viewportEl Элемент, определяющий прямоугольник в который нужно вписать тултип.
 */
export const placeTooltip = (
  tooltipEl: HTMLElement,
  holderEl: HTMLElement,
  viewportEl: HTMLElement,
) => {
  const area = getInnerRect(viewportEl);
  const offset = { x: 0, y: 0 };

  PlaceOffset.reset(tooltipEl);

  if (CanPlace.onRight(holderEl, tooltipEl, area)) {
    offset.x = PlaceOffset.onRight(tooltipEl, holderEl);
    offset.y = CorrectOffset.vertically(tooltipEl, holderEl, area);
  } else if (CanPlace.onLeft(holderEl, tooltipEl, area)) {
    offset.x = PlaceOffset.onLeft(tooltipEl, holderEl);
    offset.y = CorrectOffset.vertically(tooltipEl, holderEl, area);
  } else if (CanPlace.onBottom(holderEl, tooltipEl, area)) {
    offset.y = PlaceOffset.onBottom(tooltipEl, holderEl);
    offset.x = CorrectOffset.horizontally(tooltipEl, area);
  } else if (CanPlace.onTop(holderEl, tooltipEl, area)) {
    offset.y = PlaceOffset.onTop(tooltipEl, holderEl);
    offset.x = CorrectOffset.horizontally(tooltipEl, area);
  } else {
    offset.x = area.left + SPACES.toArea;
    offset.y = area.top + SPACES.toArea;
    tooltipEl.style.maxHeight = `${area.height - SPACES.toArea * 2}px`;
  }

  // вместо top/left/bottom/right используем translate3d, т.к. его расчет использует графический ускоритель
  tooltipEl.style.transform = `translate3d(${offset.x}px, ${offset.y}px, 0)`;
};

/**
 * Возвращает внутреннюю доступную область блока (с учетом прокруток).
 * @param element Элемент.
 * @return Прямоугольник.
 */
export const getInnerRect = (element: Element): DOMRectLike => {
  const bounds = boundsOf(element);

  const area: DOMRectLike = {
    top: bounds.top,
    left: bounds.left,
    bottom: bounds.bottom,
    right: bounds.right,
    width: bounds.width,
    height: bounds.height,
  };

  if (element.scrollWidth > area.width) {
    area.left = area.left - element.scrollLeft;
    area.width = element.scrollWidth;
    area.right = area.left + area.width;
  }

  if (element.scrollHeight > area.height) {
    area.top = area.top - element.scrollTop;
    area.height = element.scrollHeight;
    area.bottom = area.top + area.height;
  }

  return area;
};

/**
 * Получив элемент, вернет смещение, с которым можно задать координаты относительно левого верхнего угла страницы.
 * @param element Элемент, должен находиться в документе.
 * @return Смещение по осям.
 */
export const getOriginCorrection = (element: HTMLElement): { x: number; y: number } => {
  const offsetParent = findOffsetParent(element);
  const scrollParent = getScrollParent(element);
  const correction = { x: window.pageXOffset, y: window.pageYOffset };

  if (offsetParent) {
    const parentRect = boundsOf(offsetParent);
    const parentStyle = getComputedStyle(offsetParent);

    correction.x = -parentRect.left;
    correction.y = -parentRect.top;

    // border-[top/left] влияют на начало координат родителя, учитываем
    correction.x -= asNumber(parentStyle.borderLeftWidth);
    correction.y -= asNumber(parentStyle.borderTopWidth);
  }

  // если offsetParent имеет собственную прокрутку - учитываем её
  if (offsetParent === scrollParent) {
    correction.x += scrollParent.scrollLeft;
    correction.y += scrollParent.scrollTop;
  }

  return correction;
};
