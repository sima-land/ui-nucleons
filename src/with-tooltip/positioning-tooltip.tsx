import { useCallback, useEffect, useRef, useContext, ReactNode, RefObject } from 'react';
import { ViewportContext } from '../context/viewport';
import { getScrollParent } from '../helpers/get-scroll-parent';
import { placeTooltip } from './utils';
import { Tooltip } from '../tooltip';
import { useLayer } from '../helpers/layer';
import { useOutsideClick } from '../hooks';
import on from '../helpers/on';

export interface PositioningTooltipProps {
  /** Содержимое. */
  children?: ReactNode;

  /** Реф с элементом, относительно которого нужно позиционироваться. */
  openerRef: RefObject<HTMLElement | undefined>;

  /** Сработает при клике за пределами или на крестик. */
  onDismiss?: () => void;
}

/**
 * Позиционируется относительно заданного элемента.
 * @param props Свойства.
 * @return Элемент.
 */
export const PositioningTooltip = ({ openerRef, children, onDismiss }: PositioningTooltipProps) => {
  const layer = useLayer();
  const tooltipRef = useRef<HTMLDivElement>(null);
  const viewportRef = useContext(ViewportContext);

  const place = useCallback(() => {
    if (tooltipRef.current && openerRef.current) {
      placeTooltip(
        tooltipRef.current,
        openerRef.current,
        viewportRef.current || getScrollParent(tooltipRef.current),
      );
      tooltipRef.current.style.opacity = '1';
    }
  }, [openerRef, viewportRef]);

  useEffect(place);

  useEffect(() => {
    const offList: Array<() => void> = [
      on(window, 'scroll', place),
      on(window, 'resize', () => requestAnimationFrame(place)),

      // у тултипа и открывашки могут быть разные scroll-родители а может быть один,
      // учитываем оба варианта и накидываем обработчик прокрутки
      ...[tooltipRef, openerRef]
        .map(ref => getScrollParent(ref.current))
        .filter(el => el !== document.body && el !== document.documentElement)
        .reduce<HTMLElement[]>((a, b) => (!a.includes(b) && a.push(b), a), [])
        .map(el => on(el, 'scroll', place)),
    ];

    return () => offList.forEach(fn => fn());
  }, []);

  useOutsideClick(tooltipRef, () => {
    onDismiss && onDismiss();
  });

  return (
    <Tooltip
      ref={tooltipRef}
      style={{
        position: 'absolute',
        top: 0,
        left: 0,
        opacity: 0,
        zIndex: layer + 1,
      }}
      onClose={onDismiss}
    >
      {children}
    </Tooltip>
  );
};
