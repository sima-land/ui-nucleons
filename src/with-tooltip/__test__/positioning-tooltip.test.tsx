import { mount, ReactWrapper } from 'enzyme';
import { render } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { PositioningTooltip } from '../positioning-tooltip';
import { placeTooltip } from '../utils';

jest.mock('../utils', () => {
  const original = jest.requireActual('../utils');

  return {
    ...original,
    __esModule: true,
    placeTooltip: jest.fn(original.placeTooltip),
  };
});

describe('PositioningTooltip', () => {
  let wrapper: ReactWrapper;

  const actualDescriptor = {
    ...Object.getOwnPropertyDescriptor(window, 'getComputedStyle'),
  };

  beforeAll(() => {
    Object.defineProperty(window, 'getComputedStyle', {
      value: (el: any) => el.__fakeStyles || actualDescriptor.value(el),
    });
  });

  afterAll(() => {
    Object.defineProperty(window, 'getComputedStyle', actualDescriptor);
  });

  afterEach(() => {
    // enzyme не выполняет unmount после завершения it() - выполняем вручную (иначе ломаются тесты)
    wrapper && wrapper.length > 0 && wrapper.unmount();
  });

  beforeEach(() => {
    (placeTooltip as any).mockClear();
  });

  it('should renders', () => {
    wrapper = mount(
      <PositioningTooltip
        openerRef={{ current: document.createElement('div') }}
        children='Test tooltip'
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle outside click', () => {
    const spy = jest.fn();

    wrapper = mount(
      <PositioningTooltip
        openerRef={{ current: document.createElement('div') }}
        children='Test tooltip'
        onDismiss={spy}
      />,
    );

    expect(spy).toHaveBeenCalledTimes(0);

    const testEvent = new MouseEvent('click');

    act(() => {
      document.documentElement.dispatchEvent(testEvent);
    });
    wrapper.update();

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should not call placeTooltip when some ref is empty', () => {
    jest.useFakeTimers('modern');

    wrapper = mount(
      <PositioningTooltip openerRef={{ current: undefined }} children='Test tooltip' />,
    );

    expect(placeTooltip).toHaveBeenCalledTimes(0);
  });

  it('should handle window "resize" event', () => {
    jest.useFakeTimers('modern');

    wrapper = mount(
      <PositioningTooltip
        openerRef={{ current: document.createElement('div') }}
        children='Test tooltip'
      />,
    );

    expect(placeTooltip).toHaveBeenCalledTimes(1);

    act(() => {
      window.dispatchEvent(new Event('resize'));
    });

    expect(placeTooltip).toHaveBeenCalledTimes(1);

    jest.advanceTimersByTime(250);

    expect(placeTooltip).toHaveBeenCalledTimes(2);
  });

  it('should handle window "scroll" event', () => {
    wrapper = mount(
      <PositioningTooltip
        openerRef={{ current: document.createElement('div') }}
        children='Test tooltip'
      />,
    );

    expect(placeTooltip).toHaveBeenCalledTimes(1);

    act(() => {
      window.dispatchEvent(new Event('scroll'));
    });

    expect(placeTooltip).toHaveBeenCalledTimes(2);
  });

  it('should call "onDismiss" on cross click', () => {
    const spy = jest.fn();

    wrapper = mount(
      <PositioningTooltip
        openerRef={{ current: document.createElement('div') }}
        children='Test tooltip'
        onDismiss={spy}
      />,
    );

    expect(spy).toHaveBeenCalledTimes(0);

    act(() => {
      wrapper.find('svg[data-testid="tooltip:cross"]').simulate('click');
    });
    wrapper.update();

    expect(spy).toHaveBeenCalledTimes(1);
  });

  it('should handle "scroll" event of custom scroll parent', () => {
    const container = document.createElement('div');
    (container as any).__fakeStyles = { overflow: 'auto' };

    document.body.append(container);

    act(() => {
      render(
        <PositioningTooltip
          openerRef={{ current: document.createElement('div') }}
          children='Test tooltip'
        />,
        container,
      );
    });

    expect(placeTooltip).toHaveBeenCalledTimes(1);

    act(() => {
      container.dispatchEvent(new Event('scroll'));
    });

    expect(placeTooltip).toHaveBeenCalledTimes(2);

    container.remove();
  });
});
