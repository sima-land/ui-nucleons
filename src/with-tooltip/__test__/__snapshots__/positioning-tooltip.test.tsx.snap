// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositioningTooltip should renders 1`] = `
<PositioningTooltip
  openerRef={
    {
      "current": <div />,
    }
  }
>
  <ForwardRef
    style={
      {
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "top": 0,
        "zIndex": 1,
      }
    }
  >
    <div
      className="root"
      data-testid="tooltip"
      style={
        {
          "left": 0,
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "zIndex": 1,
        }
      }
    >
      <ForwardRef(CrossSVG)
        className="cross"
        data-testid="tooltip:cross"
      >
        <svg
          className="cross"
          data-testid="tooltip:cross"
          height={16}
          viewBox="0 0 16 16"
          width={16}
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M12.707 4.707a1 1 0 0 0-1.414-1.414L8 6.586 4.707 3.293a1 1 0 0 0-1.414 1.414L6.586 8l-3.293 3.293a1 1 0 1 0 1.414 1.414L8 9.414l3.293 3.293a1 1 0 0 0 1.414-1.414L9.414 8l3.293-3.293Z"
          />
        </svg>
      </ForwardRef(CrossSVG)>
      <div
        className="content"
      >
        Test tooltip
      </div>
    </div>
  </ForwardRef>
</PositioningTooltip>
`;
